test/Test/Toml/Parser/Integer.hs

```diff
@@ -81,3 +81,27 @@ integerSpecs = describe "integerP" $ do
         it "can parse numbers when hex digits are in both lowercase and uppercase" $ do
             parseInteger "0xAbCdEf" 0xAbCdEf
             parseInteger "0xaBcDeF" 0xaBcDeF
+        it "can parse numbers with underscore in hexadecimal representation" $ do
+            parseInteger "0xAb_Cd_Ef" 0xabcdef
+            parseInteger "0xA_bcd_ef" 0xabcdef
+            parseInteger "0x123_abc" 0x123abc
+            parseInteger "0xa_b_c_1_2_3" 0xabc123
+        it "can't parse when underscore is between hexadecimal prefix and suffix" $ do
+            parseInteger "0x_Abab_ca" 0
+            parseInteger "0x_ababbac" 0
+        it "can parse numbers with underscore in octal representation" $ do
+            parseInteger "0o12_34_56" 0o123456
+            parseInteger "0o1_2345_6" 0o123456
+            parseInteger "0o76_54_21" 0o765421
+            parseInteger "0o4_5_3_2_6" 0o45326
+        it "can't parse when underscore is between octal prefix and suffix" $ do
+            parseInteger "0o_123_4567" 0
+            parseInteger "0o_1234567" 0
```

Comment:
Additionally, for each such failure case, you can check `_` not followed by any numbers. Just `0b_`, `0x_` and `0o_`